Broken unit tests:  (clip_vectors_to_wbd_test & usgs_gage_crosswalk_test)

## Current behavior

Two unit tests are failing:
`clip_vectors_to_wbd_test.py` - Straightforward fix, add two new parameters. 
`usgs_gage_crosswalk_test.py` - This seems to stem from updating the shapely library.

## Expected behavior

All unit tests should pass. 

## Steps to replicate behavior (include URLs)

Clone `dev` branch. Run docker container, and issue `pytest foss_fim/unit_tests`.

Here is a condensed stacktrace from  `pytest foss_fim/unit_tests/usgs_gage_crosswalk_test.py`:

```
foss_fim/unit_tests/usgs_gage_crosswalk_test.py:49: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
foss_fim/src/usgs_gage_crosswalk.py:54: in run_crosswalk
    self.snap_to_dem_derived_flows(input_flows_filename)
foss_fim/src/usgs_gage_crosswalk.py:84: in snap_to_dem_derived_flows
    self.gages['geometry_snapped'], self.gages['snap_distance'] = self.gages.apply(self.snap_to_line, axis=1,result_type='expand').T.values
usr/local/lib/python3.8/dist-packages/geopandas/geodataframe.py:1482: in apply
    result = super().apply(
usr/local/lib/python3.8/dist-packages/pandas/core/frame.py:7541: in apply
    return op.get_result()
usr/local/lib/python3.8/dist-packages/pandas/core/apply.py:180: in get_result
    return self.apply_standard()
usr/local/lib/python3.8/dist-packages/pandas/core/apply.py:255: in apply_standard
    results, res_index = self.apply_series_generator()
usr/local/lib/python3.8/dist-packages/pandas/core/apply.py:284: in apply_series_generator
    results[i] = self.f(v)
foss_fim/src/usgs_gage_crosswalk.py:111: in snap_to_line
    snap_geom = row.geometry_ln.interpolate(row.geometry_ln.project(row.geometry))
usr/local/lib/python3.8/dist-packages/shapely/geometry/base.py:827: in project
    return shapely.line_locate_point(self, other, normalized=normalized)
usr/local/lib/python3.8/dist-packages/shapely/decorators.py:77: in wrapped
    return func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

line = <LINESTRING (1738064.409 2383699.282, 1738064.409 2383699.282, 1738064.409 2...>, other = <POINT (1738241.809 2383751.922)>, normalized = False, kwargs = {}

    @multithreading_enabled
    def line_locate_point(line, other, normalized=False, **kwargs):
        """Returns the distance to the line origin of given point.
    
        If given point does not intersect with the line, the point will first be
        projected onto the line after which the distance is taken.
    
        Parameters
        ----------
        line : Geometry or array_like
        point : Geometry or array_like
        normalized : bool, default False
            If True, the distance is a fraction of the total
            line length instead of the absolute distance.
        **kwargs
            For other keyword-only arguments, see the
            `NumPy ufunc docs <https://numpy.org/doc/stable/reference/ufuncs.html#ufuncs-kwargs>`_.
    
        Examples
        --------
        >>> from shapely import LineString, Point
        >>> line = LineString([(0, 2), (0, 10)])
        >>> point = Point(4, 4)
        >>> line_locate_point(line, point)
        2.0
        >>> line_locate_point(line, point, normalized=True)
        0.25
        >>> line_locate_point(line, Point(0, 18))
        8.0
        >>> line_locate_point(LineString(), point)
        nan
        """
        if normalized:
            return lib.line_locate_point_normalized(line, other)
        else:
>           return lib.line_locate_point(line, other)
E           RuntimeWarning: invalid value encountered in line_locate_point

usr/local/lib/python3.8/dist-packages/shapely/linear.py:90: RuntimeWarning
====================================================================================================== short test summary info =======================================================================================================
FAILED foss_fim/unit_tests/usgs_gage_crosswalk_test.py::test_usgs_gage_crosswalk::test_GageCrosswalk_success - RuntimeWarning: invalid value encountered in line_locate_point
========================================================================================================= 1 failed in 2.33s ==========================================================================================================
```

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -2,7 +2,7 @@ All notable changes to this project will be documented in this file.
 We follow the [Semantic Versioning 2.0.0](http://semver.org/) format.
 
 
-## v4.3.x.x - 2023-04-xx - [PR#893](https://github.com/NOAA-OWP/inundation-mapping/pull/893)
+## v4.3.9.1 - 2023-05-12 - [PR#893](https://github.com/NOAA-OWP/inundation-mapping/pull/893)
 
 Fix existing unit tests, remove unwanted behavior in `check_unit_errors_test.py`, update `unit_tests/README.md`
 

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -2,6 +2,22 @@ All notable changes to this project will be documented in this file.
 We follow the [Semantic Versioning 2.0.0](http://semver.org/) format.
 
 
+## v4.3.x.x - 2023-04-xx - [PR#893](https://github.com/NOAA-OWP/inundation-mapping/pull/893)
+
+Fix existing unit tests, remove unwanted behavior in `check_unit_errors_test.py`, update `unit_tests/README.md`
+
+### Changes  
+
+- `unit_tests/`
+    - `README.md` : Split up headings for setting up unit tests/running unit tests & re-formatted code block.
+    - `check_unit_errors_test.py`: Fixed unwanted behavior of test leaving behind `sample_n.txt` files in `unit_errors/`
+    - `clip_vectors_to_wbd_params.json`: Update parameters
+    - `clip_vectors_to_wbd_test.py`: Update arguments
+    - `pyproject.toml`: Ignore RuntimeWarning, to suppress pytest failure. 
+    - `usgs_gage_crosswalk_test.py`: Enhance readability of arguments in `gage_crosswalk.run_crosswalk` call
+
+<br/><br/>
+
 ## v4.3.9.0 - 2023-04-19 - [PR#889](https://github.com/NOAA-OWP/inundation-mapping/pull/889)
 
 Updates GDAL in base Docker image from 3.1.2 to 3.4.3 and updates all Python packages to latest versions, including Pandas v.2.0.0. Fixes resulting errors caused by deprecation and/or other changes in dependencies. 

--- a/unit_tests/README.md
+++ b/unit_tests/README.md
@@ -1,4 +1,4 @@
-## Inundation Mapping: Flood Inundation Mapping for U.S. National Water Model
+# Inundation Mapping: Flood Inundation Mapping for U.S. National Water Model
 
 Flood inundation mapping software configured to work with the U.S. National Water Model operated and maintained by the National Oceanic and Atmospheric Administration (NOAA) National Water Center (NWC).
 
@@ -27,7 +27,7 @@ One way is to use the incoming arg parser. Most python files include the code bl
 * Find/replace all single quotes to double quotes then cleanup the left tab formatting.
 
 
-## Running unit tests
+## Setting up unit test data
 
 Start a docker container as you normally would for any development. 
 ```bash 
@@ -62,6 +62,9 @@ If you need to run inundation tests, fun the following:
 python3 foss_fim/tools/synthesize_test_cases.py -c DEV -v fim_unit_test_data_do_not_remove \
 	-jh 1 -jb 1 -m /data/outputs/fim_unit_test_data_do_not_remove/alpha_test_metrics.csv -o
 ```
+
+## Running unit tests
+
 ### If you'd like to test the whole unit test suite:
 ```
 pytest /foss_fim/unit_tests
@@ -103,15 +106,17 @@ If one test case is choosen, it will scan all of the test files, and scan for th
 
 9) Sometimes you may want to run a full successful "happy path" version through `fim_pipeline.sh` (or similar), to get all of the files you need in place to do your testing. However, you will want to ensure that none of the outputs are being deleted during the test. One way to solve this is to put in an invalid value for the `-d` parameter (denylist). 
 ie:
-```bash
-fim_pipeline.sh -n fim_unit_test_data_do_not_remove -u 05030104 \
-	-c /foss_fim/config/params_template.env -j 1 -d /foss_fim/config/deny_unit_default.lst -o
-```
-but ours would be:
-```bash 
-fim_pipeline.sh -n fim_unit_test_data_do_not_remove -u 05030104 \
-	-c /foss_fim/config/params_template.env -j 1 -d no_list -o
-```
+	```bash
+	fim_pipeline.sh -n fim_unit_test_data_do_not_remove -u 05030104 \
+		-c /foss_fim/config/params_template.env -j 1 -d /foss_fim/config/deny_unit_default.lst -o
+	```
+	
+	 but ours would be:
+	
+	```bash 
+	fim_pipeline.sh -n fim_unit_test_data_do_not_remove -u 05030104 \
+		-c /foss_fim/config/params_template.env -j 1 -d no_list -o
+	```
 
 ## [Pytest](https://docs.pytest.org/en/7.2.x/) particulars
 

--- a/unit_tests/README.md
+++ b/unit_tests/README.md
@@ -40,7 +40,8 @@ Example:
 docker run --rm -it --name mytest \
 	-v /home/abcd/projects/dev/innudation-mapping/:/foss_fim \
 	-v /abcd_share/foss_fim/outputs/:/outputs \
-	-v /abcs_share/foss_fim/:/data \
+	-v /abcd_share/foss_fim/:/data \
+	-v /abcd_share/foss_fim/outputs_temp/:/fim_temp \
 	fim_4:dev_20220208_8eba0ee
 ```
 

--- a/unit_tests/check_unit_errors_test.py
+++ b/unit_tests/check_unit_errors_test.py
@@ -48,17 +48,16 @@ class test_check_unit_errors(unittest.TestCase):
         params = self.params["valid_data"].copy()
         
         num_dummy_files_reqd = UNIT_ERRORS_MIN_NUMBER_THRESHOLD - 1
-        ue_folder_existed = self.__create_temp_unit_errors_folder_files(params["fim_dir"], 
-                                                                        num_dummy_files_reqd)
+
+        self.__create_temp_unit_errors_folder_files(params["fim_dir"], num_dummy_files_reqd)
+
         expected_output = 0
         actual_output = src.check_unit_errors(params["fim_dir"], num_dummy_files_reqd)
         
         err_msg = "Number of dummy files IS NOT less than the overall min number of error files."
         assert expected_output == actual_output, err_msg
         
-        if (ue_folder_existed):
-            self.__remove_temp_unit_errors_folder(params["fim_dir"])
-        
+        self.__remove_temp_unit_errors_folder(params["fim_dir"])
             
     def test_check_unit_errors_fail_above_min_errors(self):
 
@@ -91,8 +90,7 @@ class test_check_unit_errors(unittest.TestCase):
         
         num_dummy_files_reqd = UNIT_ERRORS_MIN_NUMBER_THRESHOLD * 2
             
-        ue_folder_existed = self.__create_temp_unit_errors_folder_files(params["fim_dir"],
-                                                                            num_dummy_files_reqd)
+        self.__create_temp_unit_errors_folder_files(params["fim_dir"], num_dummy_files_reqd)
             
         num_total_units = math.trunc(num_dummy_files_reqd * (100 / UNIT_ERRORS_MIN_PERCENT_THRESHOLD)) + 1
         expected_output = 0
@@ -101,10 +99,7 @@ class test_check_unit_errors(unittest.TestCase):
         err_msg = "Number of dummy files IS NOT more than the overall min number of error files"
         assert expected_output == actual_output, err_msg
             
-        if (ue_folder_existed):
-            self.__remove_temp_unit_errors_folder(params["fim_dir"]) 
-            
-        # We have to put the unit_errors folders back to the way is was.
+        # We have to put the unit_errors folders back to the way is was.           
         self.__remove_temp_unit_errors_folder(params["fim_dir"])
 
 

--- a/unit_tests/clip_vectors_to_wbd_params.json
+++ b/unit_tests/clip_vectors_to_wbd_params.json
@@ -12,11 +12,13 @@
 		"nwm_catchments": "/data/inputs/nwm_hydrofabric/nwm_catchments.gpkg",
 		"subset_nwm_catchments": "/data/outputs/fim_unit_test_data_do_not_remove/05030104/nwm_catchments_proj_subset.gpkg",
 		"nld_lines": "/data/inputs/nld_vectors/huc2_levee_lines/nld_preprocessed_05.gpkg",
+		"nld_lines_preprocessed": "/data/inputs/nld_vectors/3d_nld_preprocessed_230314.gpkg",
 		"landsea": "/data/inputs/landsea/water_polygons_us.gpkg",
 		"nwm_streams": "/data/inputs/nwm_hydrofabric/nwm_flows.gpkg",
 		"subset_landsea": "/data/outputs/fim_unit_test_data_do_not_remove/05030104/LandSea_subset.gpkg",
 		"nwm_headwaters": "/data/inputs/nhdplus_vectors_aggregate/agg_nhd_headwaters_adj.gpkg",
 		"subset_nld_lines": "/data/outputs/fim_unit_test_data_do_not_remove/05030104/nld_subset_levees.gpkg",
+		"subset_nld_lines_preprocessed": "/data/outputs/fim_unit_test_data_do_not_remove/05030104/3d_nld_subset_levees_burned.gpkg",
 		"wbd_buffer_distance": 5000,
 		"levee_protected_areas": "/data/inputs/nld_vectors/Levee_protected_areas.gpkg",
 		"subset_levee_protected_areas": "/data/outputs/fim_unit_test_data_do_not_remove/05030104/LeveeProtectedAreas_subset.gpkg"

--- a/unit_tests/clip_vectors_to_wbd_test.py
+++ b/unit_tests/clip_vectors_to_wbd_test.py
@@ -52,11 +52,13 @@ class test_clip_vectors_to_wbd(unittest.TestCase):
                                      nwm_catchments = params["nwm_catchments"],
                                      subset_nwm_catchments = params["subset_nwm_catchments"],
                                      nld_lines = params["nld_lines"],
+                                     nld_lines_preprocessed = params["nld_lines_preprocessed"],
                                      landsea = params["landsea"],
                                      nwm_streams = params["nwm_streams"],
                                      subset_landsea = params["subset_landsea"],
                                      nwm_headwaters = params["nwm_headwaters"],
                                      subset_nld_lines = params["subset_nld_lines"],
+                                     subset_nld_lines_preprocessed = params["subset_nld_lines_preprocessed"],
                                      wbd_buffer_distance = params["wbd_buffer_distance"],
                                      levee_protected_areas = params["levee_protected_areas"],
                                      subset_levee_protected_areas = params["subset_levee_protected_areas"])

--- a/unit_tests/pyproject.toml
+++ b/unit_tests/pyproject.toml
@@ -3,5 +3,6 @@ filterwarnings = [
     "error",
     "ignore::UserWarning",
     "ignore::DeprecationWarning",
-    "ignore::PendingDeprecationWarning"
+    "ignore::PendingDeprecationWarning",
+    "ignore::RuntimeWarning"
 ]

--- a/unit_tests/usgs_gage_crosswalk_test.py
+++ b/unit_tests/usgs_gage_crosswalk_test.py
@@ -46,8 +46,13 @@ class test_usgs_gage_crosswalk(unittest.TestCase):
         gage_crosswalk = GageCrosswalk(params["usgs_gages_filename"], params["branch_id"])
 
         # Run crosswalk
-        gage_crosswalk.run_crosswalk(params["input_catchment_filename"], params["input_flows_filename"], 
-                            params["dem_filename"], params["dem_adj_filename"], params["output_table_filename"])
+        # gage_crosswalk.run_crosswalk(params["input_catchment_filename"], params["input_flows_filename"], 
+        #                     params["dem_filename"], params["dem_adj_filename"], params["output_table_filename"])
+        gage_crosswalk.run_crosswalk(params["input_catchment_filename"], \
+                                     params["input_flows_filename"], \
+                                     params["dem_filename"], \
+                                     params["dem_adj_filename"], \
+                                     params["output_table_filename"])
 
         # Make sure that the usgs_elev_table.csv was written
         msg = f'{params["output_table_filename"]} does not exist'

--- a/unit_tests/usgs_gage_crosswalk_test.py
+++ b/unit_tests/usgs_gage_crosswalk_test.py
@@ -46,8 +46,6 @@ class test_usgs_gage_crosswalk(unittest.TestCase):
         gage_crosswalk = GageCrosswalk(params["usgs_gages_filename"], params["branch_id"])
 
         # Run crosswalk
-        # gage_crosswalk.run_crosswalk(params["input_catchment_filename"], params["input_flows_filename"], 
-        #                     params["dem_filename"], params["dem_adj_filename"], params["output_table_filename"])
         gage_crosswalk.run_crosswalk(params["input_catchment_filename"], \
                                      params["input_flows_filename"], \
                                      params["dem_filename"], \

